### ISSUE TITLE
How to create a new pdf from a list of PdfPage?

### ISSUE THREAD

**Issue by Dragon2fly** [pypdfium2-team/pypdfium2#205]
I see that there is a `PdfDocument.import_pages(pdf, pages=None, index=None)` method used to add pages from another pdf file.
But it uses `pages` as index numbers, not a collection of `PdfPage` objects.

Because the `PdfPage` object doesn't contain any information about its page number,
so if I already have a list of `PdfPage` objects after reading a pdf, how can I create a new pdf from it?

**Comment by mara004** [comment]
AFAIK, pdfium's C API does not provide an object-based page importer.
And emulating that in the support model would be a bad idea because page indices can change.
Tracking that is a wonky business, and particularly it's not possible to track changes made using the raw API.
(See [this thread](https://github.com/ajrcarey/pdfium-render/issues/67#issuecomment-1481446366) for background reading if you're curious.)

So, just design your code to import pages based on indices. If you do not change page order, you could just store tuples of `(obj, index)`, or assign an `index` attribute to the `PdfPage` objects on after creation.

**Comment by mara004** [comment]
Sounds seriously inelegant. I'd strongly recommend that you work with page indices instead.
If you don't change page order of the input documents, that shouldn't be any problem at all.

If you do, then just make sure you track all move/delete/insert operations and change indices accordingly.
While we cannot implement that in the support model API because of the raw API thing, it should be quite doable as a caller encapsulating pdfium, where you are in charge of all page order operations.

**Comment by mara004** [comment]
Thinking about this a bit more, I guess we actually could add an `index` attribute to `PdfPage` objects, provided we point out that the caller is responsible for changing the value (or not using it) if modifying page order using the raw API.

**Comment by mara004** [comment]
Summarized this for tasks.md as follows:

> Consider adding a page object index property backed by a view of released pages on document level (`list[None | weakref.ReferenceType[PdfPage]]`) that is updated on page order changes, despite a certain risk that the page view might not match the actual state if changes are not tracked properly. (To implement this safely, PDFium would need to provide a page indexing function natively)

**Comment by Dragon2fly** [comment]
Hi @mara004 

Thanks for considering tracking the page number.
In the meantime, I use the below code to circumvent the page number.
The only weak point that I found is, exporting the pdf from a collection of PdfPage has its file size larger than from using the collection of the index, proportionally to the number of pages. It seems that the Pdf is saved incrementally.

```python
from typing import Union, BinaryIO
from pathlib import Path
from io import BytesIO

import pypdfium2 as pdfium


class PdfPage:
    """
    wrapper class of pdfium.PdfPage to be able to add split/merge/export freely
    with just a collection of pdfium.PdfPage and don't need to track the index
    """

    def __init__(self, document: pdfium.PdfDocument, page_no: int):
        self._doc = pdfium.PdfDocument.new()
        self._doc.import_pages(document, [page_no])
        self._data = self._doc[0]
        self._text_page = None

    def show(self, scale=2):
        bitmap = self._data.render(
            scale=scale,   # 72dpi * scale resolution
            rotation=0,    # no additional rotation
        )

        pil_image = bitmap.to_pil()
        pil_image.show()

    @property
    def text_page(self):
        if not self._text_page:
            self._text_page = self._data.get_textpage()
        return self._text_page

    def extract_text(self, bbox: tuple = None):
        bounding_box = bbox
        if bounding_box:
            left, bottom, right, top = bounding_box
            self.text_page.get_text_bounded(left, bottom, right, top)
        else:
            return self.text_page.get_text_range()

    def save_changes(self, ):
        """Save changed to the page content to the doc before export to file"""
        self._data.gen_content()

    def export(self, filename: Union[str, BinaryIO, Path], *args, **kwargs):
        self._doc.save(filename)

    def __getattr__(self, item):
        return getattr(self._data, item)


class PdfDocument(pdfium.PdfDocument):
    def get_page(self, index):
        return PdfPage(self, index)
    
    def as_bytes_io(self, pages: Union[list[PdfPage], list[int]] = None):
        stream = BytesIO()

        if not pages:
            # save current pdf into stream
            self.save(stream, version=17)
            return stream
        else:
            # create a new pdf with a collected pages
            doc = self.new()
            if type(pages[0]) is int:
                # It must be a list of zero-based index
                doc.import_pages(self, pages)
            else:
                # list of PdfPage object
                for p in pages:
                    doc.import_pages(p.doc)
            doc.save(stream, version=17)
            return stream

    def export(self, to_filename: Union[str, Path], pages: Union[list[PdfPage], list[int]] = None):
        """ Save pdf to file-like object
            Overwrite if to_filename already exists
        """
        byte_stream = self.as_bytes_io(pages)
        with open(to_filename, 'wb') as fo:
            fo.write(byte_stream.getbuffer())
```

**Comment by mara004** [comment]
Isolating pages as separate in-memory documents is bound to cause overhead in all respects (memory load / performance / file size). I *strongly recommend* that you rewrite your code with indices as hinted above.

Can you tell if you are doing any page order modifications in source documents? If not, using indices should be really simple, much simpler than your code above.

**Comment by Dragon2fly** [comment]
Thank you for your feedback. I'll consider adding the page index.

I do change the pages' order and then save them into the **new** pdf file.
I don't delete pages, just take the pages that I need from multiple pdf files into a new one.
So I guess I'm safe with the indices thing. 

The code of `PdfPage` above mostly hide the complexity of getting the text and shows a page as an image for debugging.
Not so much thing to do with in-memory documents.

**Comment by mara004** [comment]
> I do change the pages' order and then save them into the new pdf file.
> I don't delete pages, just take the pages that I need from multiple pdf files into a new one.
> So I guess I'm safe with the indices thing.

Yes, as long as you don't do any in-place page order modification but only copy pages into new document handles, indices are fine. That will avoid some unnecessary processing and slightly simplify your code.

> The code of PdfPage above mostly hide the complexity of getting the text and shows a page as an image for debugging.
> Not so much thing to do with in-memory documents.

Ah, sorry, I was under the (wrong) impression that there was an additional layer of "export to bytes & reopen" on init of your PdfPage, but obviously we can just import from the existing document handle, so less bad than I initially thought.
